File: fretecalc/app/src/main/java/com/example/fretecalc/viewmodels/SearchActivityViewModel.java
package com.example.fretecalc.viewmodels;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.fretecalc.data.remote.PlacesService;
import com.google.android.libraries.places.api.model.AutocompletePrediction;
import com.google.android.libraries.places.api.model.Place;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

import static com.example.fretecalc.utils.ApisConstants.GOOGLE_API_KEY;

public class SearchActivityViewModel extends AndroidViewModel {
    private MutableLiveData<List<AutocompletePrediction>> predictionList = new MutableLiveData<>();
    private MutableLiveData<Place> place = new MutableLiveData<>();
    private CompositeDisposable disposable = new CompositeDisposable();
    private PlacesService placesService = new PlacesService(getApplication().getApplicationContext(), GOOGLE_API_KEY);
    private MutableLiveData<String> error = new MutableLiveData<>();

    public SearchActivityViewModel(@NonNull Application application) {
        super(application);
    }

    public LiveData<List<AutocompletePrediction>> getPredictionList() {
        return this.predictionList;
    }

    public LiveData<String> getError() {
        return this.error;
    }

    public LiveData<Place> getPlace() {
        return this.place;
    }

    public void getPredictions(CharSequence query) {
        disposable.add(
                placesService.getPredictions(query)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(autocompletePredictions -> {
                            predictionList.setValue(autocompletePredictions);
                        }, throwable -> {
                            error.setValue(throwable.getMessage());
                        })
        );
    }

    public void getPlaceById(String placeId) {
        disposable.add(
                placesService.getPlaceById(placeId)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(place1 -> {
                            place.setValue(place1);
                        }, throwable -> {
                            error.setValue(throwable.getMessage());
                        })
        );
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        disposable.clear();
    }
}

File: FreightPad/app/src/main/java/dev/khalil/freightpad/ui/viewModel/SearchActivityViewModel.kt
package dev.khalil.freightpad.ui.viewModel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import dev.khalil.freightpad.api.repository.SearchApiRepository
import dev.khalil.freightpad.model.Place
import io.reactivex.BackpressureStrategy
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject
import java.util.concurrent.TimeUnit

class SearchActivityViewModel(private val repository: SearchApiRepository) : ViewModel() {

  private val searchResultLiveData = MutableLiveData<List<Place>>()
  val searchResult: LiveData<List<Place>>
    get() = searchResultLiveData

  private val errorMutableLiveData = MutableLiveData<Boolean>()
  val error: LiveData<Boolean>
    get() = errorMutableLiveData

  private val loadingMutableLiveData = MutableLiveData<Boolean>()
  val loading: LiveData<Boolean>
    get() = loadingMutableLiveData

  private val _searchInput = BehaviorSubject.create<String>()
  private val searchInput = _searchInput.toFlowable(BackpressureStrategy.LATEST)

  private val compositeDisposable = CompositeDisposable()
  private var lastSearch = ""

  init {
    compositeDisposable.add(searchInput.debounce(500, TimeUnit.MILLISECONDS)
      .subscribeOn(Schedulers.io())
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe { executeSearch(it) }
    )

    errorMutableLiveData.value = false
    loadingMutableLiveData.value = false
  }

  fun search(query: String) {
    if (query.length <= 3)
      return

    _searchInput.onNext(query)
  }

  fun onDestroy() {
    compositeDisposable.dispose()
    compositeDisposable.clear()
  }

  private fun executeSearch(query: String) {
    if (lastSearch == query)
      return

    compositeDisposable.add(
      repository.searchLocation(query)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .doOnSubscribe {
          loadingMutableLiveData.value = true
          errorMutableLiveData.value = false
        }
        .doAfterTerminate { loadingMutableLiveData.value = false }
        .subscribe({ responseList ->
          loadingMutableLiveData.value = false
          searchResultLiveData.value = responseList
          lastSearch = query
        }, {
          lastSearch = ""
          errorMutableLiveData.value = true
        }))
  }
}
